thirdparty/cmdconfig/commands/cmdeval/cmdeval.go

```diff
@@ -102,14 +109,71 @@ type EvalFnRunner struct {
 	IncludeMetaResources bool
 	Ctx                  context.Context
 	Selector             kptfile.Selector
+	dataItems            []string
 }
 
 func (r *EvalFnRunner) runE(c *cobra.Command, _ []string) error {
 	err := runner.HandleError(r.Ctx, r.RunFns.Execute())
 	if err != nil {
 		return err
 	}
-	return cmdutil.WriteFnOutput(r.Dest, r.OutContent.String(), r.FromStdin, printer.FromContextOrDie(r.Ctx).OutStream())
+	if err = cmdutil.WriteFnOutput(r.Dest, r.OutContent.String(), r.FromStdin,
+		printer.FromContextOrDie(r.Ctx).OutStream()); err != nil {
+		return err
+	}
+	if r.AddImage {
+		r.AddImageToKptfile()
+	}
+	return nil
+}
+
+// AddImageToKptfile adds the function image and config to Kptfile
+func (r *EvalFnRunner) AddImageToKptfile() {
+	pr := printer.FromContextOrDie(r.Ctx)
+	kptFile, err := pkg.ReadKptfile(filesys.FileSystemOrOnDisk{}, r.Dest)
+	if err != nil {
+		pr.Printf("image not added: Kptfile not exists\n")
+		return
+	}
+	// TODO(yuwenma): Right now we cannot tell a function is a mutator or validator. Once kpt supports
+	// OCI images, we can add annotations to image and find out the function type from these annotations.
+	if kptFile.Pipeline == nil {
+		kptFile.Pipeline = &kptfile.Pipeline{}
+	}
+	if kptFile.Pipeline.Mutators == nil {
+		kptFile.Pipeline.Mutators = []kptfile.Function{}
+	} else {
+		for _, m := range kptFile.Pipeline.Mutators {
+			if m.Name == r.Image || m.Image == r.Image {
+				pr.Printf("skip adding image: already exists in Kptfile\n")
+				return
+			}
+		}
+	}
+	var newMutator kptfile.Function
+	if r.FnConfigPath != "" {
+		newMutator = kptfile.Function{Name: r.Image, Image: r.Image, ConfigPath: r.FnConfigPath}
+	} else {
+		data := map[string]string{}
+		for i, s := range r.dataItems {
+			kv := strings.SplitN(s, "=", 2)
+			if i == 0 && len(kv) == 1 {
+				continue
+			}
+			data[kv[0]] = kv[1]
+		}
+		if len(data) == 0 {
+			newMutator = kptfile.Function{Name: r.Image, Image: r.Image}
+		} else {
+			newMutator = kptfile.Function{Name: r.Image, Image: r.Image, ConfigMap: data}
+		}
+	}
+	kptFile.Pipeline.Mutators = append(kptFile.Pipeline.Mutators, newMutator)
+	if err = kptfileutil.WriteFile(r.Dest, kptFile); err != nil {
+		pr.Printf("image is not added to Kptfile: %v\n", err)
+		return
+	}
```

Comment:
I would suggest splitting this function in two parts (two separate smaller functions):

1. First part converts a given `fn eval` invocation into `kptfile.Function`
2. Second part takes a `kptfile.Function` pointer and saves it to a given `kptFile` struct

Reply:
done.